doc/HLD_of_FOL.rst

```diff
@@ -0,0 +1,290 @@
+
+- [R.FOL.ORDERING]: A FOL records are applied in order. A FOL record contains enough information to detect when all necessary pre-requisite state changes have been applied;
+
+- [R.FOL.DEPENDENCIES]: Operation dependencies can be discovered through FOL. FOL contains enough information to determine dependencies between operations;
+
+- [R.FOL.DIX]: FOL supports DIX;
+
+- [R.FOL.SNS]: FOL supports SNS;
+
+- [R.FOL.REINT]: FOL can be used for cache reintegration. FOL contains enough information to find out what has to be re-integrated;
+
+- [R.FOL.PRUNE]: FOL can be pruned. A mechanism exists to determine what portions of FOL can be re-claimed;
+
+- [R.FOL.REPLAY]: FOL records can be replayed;
+
+- [R.FOL.REDO]: FOL can be used for redo-only recovery;
+
+- [R.FOL.UNDO]: FOL can be used for undo-redo recovery;
+
+- [R.FOL.EPOCHS]: FOL records for a given epoch can be found efficiently;
+
+- [R.FOL.CONSUME.SYNC]: storage applications can process FOL records synchronously;
+
+- [R.FOL.CONSUME.ASYNC]: storage applications can process FOL records asynchronously;
+
+- [R.FOL.CONSUME.RESUME]: a storage application can be resumed after a failure;
+
+- [R.FOL.ADDB]: FOL is integrated with ADDB. ADDB records matching a given FOL record can be found efficiently;
+
+- [R.FOL.FILE]: FOL records pertaining to a given file(-set) can be found efficiently.
+
+******************
+Design Highlights
+******************
+
+A fol record is identified by its LSN. LSN are defined and selected as to be able to encode various partial orders imposed on fol records by the requirements.
+
+**************************
+Functional Specification
+**************************
+
+The fol manager exports two interfaces:
+
+- main interface used by the request handler. Through this interface fol records can be added to the fol and the fol can be forced (i.e., made persistent up to a certain record);
+
+- auxiliary interfaces, used for fol pruning and querying.
+
+***********************
+Logical Specification
+***********************
+
+Overview
+=========
+
+Fol is stored in a transactional container1 populated with records indexed2 by lsn. An lsn is used to refer to a point in fol from other meta-data tables (epochs table, object index, sessions table, etc.). To make such references more flexible, a fol, in addition to genuine records corresponding to updates, might contain pseudo-records marking points on interest in the fol to which other file system tables might want to refer to (for example, an epoch boundary, a snapshot origin, a new server secret key, etc.). By abuse of terminology, such pseudo-records will be called fol records too. Similarly, as part of redo-recovery implementation, DTM might populate a node fol with records describing updates to be performed on other nodes.
```

Comment:
Please note the first sentence:
```
Fol is stored in a transactional container1 populated with records indexed2 by lsn
```

There are two foot notes. Please arrange them in correct format.

Reply:
the word "lsn" should be converted to "LSN".

Reply:
"a fol" should be changed to "an FOL", just like "an LSN".

Reply:
Incorporated.